File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala
package io.unitycatalog.connectors.spark

import com.google.gson.internal.bind.DefaultDateTypeAdapter.DateType
import io.unitycatalog.client.{ApiClient, ApiException}
import io.unitycatalog.client.api.{SchemasApi, TablesApi, TemporaryTableCredentialsApi}
import io.unitycatalog.client.model.{AwsCredentials, CreateSchema, GenerateTemporaryTableCredential, ListTablesResponse, SchemaInfo, TableOperation, TableType, UpdateSchema}
import org.apache.arrow.vector.types.pojo.ArrowType.Timestamp

import java.net.URI
import java.util
import org.apache.spark.sql.catalyst.TableIdentifier
import org.apache.spark.sql.catalyst.analysis.{NoSuchNamespaceException, NoSuchTableException}
import org.apache.spark.sql.catalyst.catalog.{CatalogStorageFormat, CatalogTable, CatalogTableType, CatalogUtils}
import org.apache.spark.sql.connector.catalog.NamespaceChange.SetProperty
import org.apache.spark.sql.connector.catalog._
import org.apache.spark.sql.connector.expressions.Transform
import org.apache.spark.sql.types.{BinaryType, BooleanType, ByteType, DataType, DoubleType, FloatType, IntegerType, LongType, ShortType, StringType, StructField, StructType, TimestampNTZType, TimestampType}
import org.apache.spark.sql.util.CaseInsensitiveStringMap

import scala.collection.convert.ImplicitConversions._
import scala.jdk.CollectionConverters._

/**
 * A Spark catalog plugin to get/manage tables in Unity Catalog.
 */
class UCSingleCatalog extends TableCatalog with SupportsNamespaces {

  private var deltaCatalog: TableCatalog = null

  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
    val proxy = new UCProxy()
    proxy.initialize(name, options)
    deltaCatalog = Class.forName("org.apache.spark.sql.delta.catalog.DeltaCatalog")
      .getDeclaredConstructor().newInstance().asInstanceOf[TableCatalog]
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].setDelegateCatalog(proxy)
  }

  override def name(): String = deltaCatalog.name()

  override def listTables(namespace: Array[String]): Array[Identifier] = deltaCatalog.listTables(namespace)

  override def loadTable(ident: Identifier): Table = deltaCatalog.loadTable(ident)

  override def tableExists(ident: Identifier): Boolean = {
    deltaCatalog.tableExists(ident)
  }

  override def createTable(
      ident: Identifier,
      columns: Array[Column],
      partitions: Array[Transform],
      properties: util.Map[String, String]): Table = {
    deltaCatalog.createTable(ident, columns, partitions, properties)
  }

  override def createTable(ident: Identifier, schema: StructType, partitions: Array[Transform], properties: util.Map[String, String]): Table = {
    throw new AssertionError("deprecated `createTable` should not be called")
  }
  override def alterTable(ident: Identifier, changes: TableChange*): Table = ???

  override def dropTable(ident: Identifier): Boolean = deltaCatalog.dropTable(ident)

  override def renameTable(oldIdent: Identifier, newIdent: Identifier): Unit = ???

  override def listNamespaces(): Array[Array[String]] = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].listNamespaces()
  }

  override def listNamespaces(namespace: Array[String]): Array[Array[String]] = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].listNamespaces(namespace)
  }

  override def loadNamespaceMetadata(namespace: Array[String]): util.Map[String, String] = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].loadNamespaceMetadata(namespace)
  }

  override def createNamespace(namespace: Array[String], metadata: util.Map[String, String]): Unit = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].createNamespace(namespace, metadata)
  }

  override def alterNamespace(namespace: Array[String], changes: NamespaceChange*): Unit = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].alterNamespace(namespace, changes: _*)
  }

  override def dropNamespace(namespace: Array[String], cascade: Boolean): Boolean = {
    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].dropNamespace(namespace, cascade)
  }
}

// An internal proxy to talk to the UC client.
private class UCProxy extends TableCatalog with SupportsNamespaces {
  private[this] var name: String = null
  private[this] var tablesApi: TablesApi = null
  private[this] var schemasApi: SchemasApi = null
  private[this] var temporaryTableCredentialsApi: TemporaryTableCredentialsApi = null

  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
    this.name = name
    val urlStr = options.get("uri")
    if (urlStr == null) {
      throw new IllegalArgumentException("uri must be specified for Unity Catalog")
    }
    val url = new URI(urlStr)
    var client = new ApiClient()
      .setHost(url.getHost)
      .setPort(url.getPort)
      .setScheme(url.getScheme)
    val token = options.get("token")
    if (token != null && token.nonEmpty) {
      client = client.setRequestInterceptor { request =>
        request.header("Authorization", "Bearer " + token)
      }
    }
    tablesApi = new TablesApi(client)
    temporaryTableCredentialsApi = new TemporaryTableCredentialsApi(client)
    schemasApi = new SchemasApi(client)
  }

  override def name(): String = {
    assert(this.name != null)
    this.name
  }

  override def listTables(namespace: Array[String]): Array[Identifier] = {
    checkUnsupportedNestedNamespace(namespace)

    val catalogName = this.name
    val schemaName = namespace.head
    val maxResults = 0
    val pageToken = null
    val response: ListTablesResponse = tablesApi.listTables(catalogName, schemaName, maxResults, pageToken)
    response.getTables.toSeq.map(table => Identifier.of(namespace, table.getName)).toArray
  }


  override def loadTable(ident: Identifier): Table = {
    val t = try {
      tablesApi.getTable(name + "." + ident.toString)
    } catch {
      case e: ApiException if e.getCode == 404 =>
        throw new NoSuchTableException(ident)
    }
    val identifier = TableIdentifier(t.getName, Some(t.getSchemaName), Some(t.getCatalogName))
    val partitionCols = scala.collection.mutable.ArrayBuffer.empty[(String, Int)]
    val fields = t.getColumns.asScala.map { col =>
      Option(col.getPartitionIndex).foreach { index =>
        partitionCols += col.getName -> index
      }
      StructField(col.getName, DataType.fromDDL(col.getTypeText), col.getNullable)
        .withComment(col.getComment)
    }.toArray
    val uri = CatalogUtils.stringToURI(t.getStorageLocation)
    val tableId = t.getTableId
    val credential: AwsCredentials = temporaryTableCredentialsApi
      .generateTemporaryTableCredentials(
        // TODO: at this time, we don't know if the table will be read or written. We should get
        //       credential in a later phase.
        new GenerateTemporaryTableCredential().tableId(tableId).operation(TableOperation.READ_WRITE)
      )
      .getAwsTempCredentials
    val extraSerdeProps = if (uri.getScheme == "s3") {
      Map(
        // TODO: how to support s3:// properly?
        "fs.s3a.access.key" -> credential.getAccessKeyId,
        "fs.s3a.secret.key" -> credential.getSecretAccessKey,
        "fs.s3a.session.token" -> credential.getSessionToken,
        "fs.s3a.path.style.access" -> "true"
      )
    } else {
      Map.empty
    }
    val sparkTable = CatalogTable(
      identifier,
      tableType = if (t.getTableType == TableType.MANAGED) {
        CatalogTableType.MANAGED
      } else {
        CatalogTableType.EXTERNAL
      },
      storage = CatalogStorageFormat.empty.copy(
        locationUri = Some(uri),
        properties = t.getProperties.asScala.toMap ++ extraSerdeProps
      ),
      schema = StructType(fields),
      provider = Some(t.getDataSourceFormat.getValue),
      createTime = t.getCreatedAt,
      tracksPartitionsInCatalog = false,
      partitionColumnNames = partitionCols.sortBy(_._2).map(_._1).toSeq
    )
    // Spark separates table lookup and data source resolution. To support Spark native data
    // sources, here we return the `V1Table` which only contains the table metadata. Spark will
    // resolve the data source and create scan node later.
    Class.forName("org.apache.spark.sql.connector.catalog.V1Table")
      .getDeclaredConstructor(classOf[CatalogTable]).newInstance(sparkTable)
      .asInstanceOf[Table]
  }

  override def createTable(ident: Identifier, schema: StructType, partitions: Array[Transform], properties: util.Map[String, String]): Table = {
    checkUnsupportedNestedNamespace(ident.namespace())
    assert(properties.get(TableCatalog.PROP_EXTERNAL) == null)
    assert(properties.get("provider") != null)

    val createTable = new CreateTable()
    createTable.setName(ident.name())
    createTable.setSchemaName(ident.namespace().head)
    createTable.setCatalogName(this.name)
    val storageLocation = properties.get(TableCatalog.PROP_LOCATION)
    if (storageLocation == null) {
      // TODO: Unity Catalog does not support managed tables now.
      throw new ApiException("Unity Catalog does not support managed table.")
    }
    createTable.setTableType(TableType.EXTERNAL)
    createTable.setStorageLocation(storageLocation)

    val columns: Seq[ColumnInfo] = schema.fields.toSeq.zipWithIndex.map { case (field, i) =>
      val column = new ColumnInfo()
      column.setName(field.name)
      if (field.getComment().isDefined) {
        column.setComment(field.getComment.get)
      }
      column.setNullable(field.nullable)
      column.setTypeText(field.dataType.simpleString)
      column.setTypeName(convertDataTypeToTypeName(field.dataType))
      column.setTypeJson(field.dataType.json)
      column.setPosition(i)
      column
    }
    createTable.setColumns(columns)
    val format: String = properties.get("provider")
    createTable.setDataSourceFormat(convertDatasourceFormat(format))
    tablesApi.createTable(createTable)
    null
  }

  private def convertDatasourceFormat(format: String): DataSourceFormat = {
    format.toUpperCase match {
      case "PARQUET" => DataSourceFormat.PARQUET
      case "CSV" => DataSourceFormat.CSV
      case "DELTA" => DataSourceFormat.DELTA
      case "JSON" => DataSourceFormat.JSON
      case "ORC" => DataSourceFormat.ORC
      case "TEXT" => DataSourceFormat.TEXT
      case "AVRO" => DataSourceFormat.AVRO
      case _ => throw new ApiException("DataSourceFormat not supported: " + format)
    }
  }

  private def convertDataTypeToTypeName(dataType: DataType): ColumnTypeName = {
    dataType match {
      case StringType => ColumnTypeName.STRING
      case BooleanType => ColumnTypeName.BOOLEAN
      case ShortType => ColumnTypeName.SHORT
      case IntegerType => ColumnTypeName.INT
      case LongType => ColumnTypeName.LONG
      case FloatType => ColumnTypeName.FLOAT
      case DoubleType => ColumnTypeName.DOUBLE
      case ByteType => ColumnTypeName.BYTE
      case BinaryType => ColumnTypeName.BINARY
      case TimestampNTZType => ColumnTypeName.TIMESTAMP_NTZ
      case TimestampType => ColumnTypeName.TIMESTAMP
      case _ => throw new ApiException("DataType not supported: " + dataType.simpleString)
    }
  }

  override def alterTable(ident: Identifier, changes: TableChange*): Table = ???

  override def dropTable(ident: Identifier): Boolean = {
    checkUnsupportedNestedNamespace(ident.namespace())
    val ret =
      tablesApi.deleteTable(Seq(this.name, ident.namespace()(0), ident.name()).mkString("."))
    if (ret == 200) true else false
  }

  override def renameTable(oldIdent: Identifier, newIdent: Identifier): Unit = ???

  private def checkUnsupportedNestedNamespace(namespace: Array[String]): Unit = {
    if (namespace.length > 1) {
      throw new ApiException("Nested namespaces are not supported:  " + namespace.mkString("."))
    }
  }

  override def listNamespaces(): Array[Array[String]] = {
    schemasApi.listSchemas(name, 0, null).getSchemas.asScala.map { schema =>
      Array(schema.getName)
    }.toArray
  }

  override def listNamespaces(namespace: Array[String]): Array[Array[String]] = {
    throw new UnsupportedOperationException("Multi-layer namespace is not supported in Unity Catalog")
  }

  override def loadNamespaceMetadata(namespace: Array[String]): util.Map[String, String] = {
    checkUnsupportedNestedNamespace(namespace)
    val schema = try {
      schemasApi.getSchema(name + "." + namespace(0))
    } catch {
      case e: ApiException if e.getCode == 404 =>
        throw new NoSuchNamespaceException(namespace)
    }
    // flatten the schema properties to a map, with the key prefixed by "properties:"
    val metadata = schema.getProperties.asScala.map {
      case (k, v) =>  SchemaInfo.JSON_PROPERTY_PROPERTIES + ":" + k -> v
    }
    metadata(SchemaInfo.JSON_PROPERTY_NAME) = schema.getName
    metadata(SchemaInfo.JSON_PROPERTY_CATALOG_NAME) = schema.getCatalogName
    metadata(SchemaInfo.JSON_PROPERTY_COMMENT) = schema.getComment
    metadata(SchemaInfo.JSON_PROPERTY_FULL_NAME) = schema.getFullName
    metadata(SchemaInfo.JSON_PROPERTY_CREATED_AT) = if (schema.getCreatedAt != null) {schema.getCreatedAt.toString} else {"null"}
    metadata(SchemaInfo.JSON_PROPERTY_UPDATED_AT) = if (schema.getUpdatedAt != null) {schema.getUpdatedAt.toString} else {"null"}
    metadata(SchemaInfo.JSON_PROPERTY_SCHEMA_ID) = schema.getSchemaId
    metadata.asJava
  }

  override def createNamespace(namespace: Array[String], metadata: util.Map[String, String]): Unit = {
    checkUnsupportedNestedNamespace(namespace)
    val createSchema = new CreateSchema()
    createSchema.setCatalogName(this.name)
    createSchema.setName(namespace.head)
    createSchema.setProperties(metadata)
    schemasApi.createSchema(createSchema)
  }

  override def alterNamespace(namespace: Array[String], changes: NamespaceChange*): Unit = {
    checkUnsupportedNestedNamespace(namespace)
    val updateSchema = new UpdateSchema()
    changes.foreach {
      case setProperty: SetProperty =>
        updateSchema.putPropertiesItem(setProperty.property(), setProperty.value())
      case p => throw new ApiException("UC does not support namespace change operation: " + p.toString)
    }
    schemasApi.updateSchema(name + "." + namespace.head, updateSchema)
  }

  override def dropNamespace(namespace: Array[String], cascade: Boolean): Boolean = {
    checkUnsupportedNestedNamespace(namespace)
    schemasApi.deleteSchema(name + "." + namespace.head, cascade)
    true
  }
}

File: connectors/spark/src/main/java/io/unitycatalog/spark/UCSingleCatalog.java
package io.unitycatalog.spark;

import io.unitycatalog.client.ApiClient;
import io.unitycatalog.client.ApiException;
import io.unitycatalog.client.api.SchemasApi;
import io.unitycatalog.client.api.TablesApi;
import io.unitycatalog.client.api.TemporaryCredentialsApi;
import io.unitycatalog.client.auth.TokenProvider;
import io.unitycatalog.client.model.CreateStagingTable;
import io.unitycatalog.client.model.GenerateTemporaryPathCredential;
import io.unitycatalog.client.model.GenerateTemporaryTableCredential;
import io.unitycatalog.client.model.PathOperation;
import io.unitycatalog.client.model.StagingTableInfo;
import io.unitycatalog.client.model.TableOperation;
import io.unitycatalog.client.model.TemporaryCredentials;
import io.unitycatalog.client.retry.JitterDelayRetryPolicy;
import io.unitycatalog.spark.auth.AuthConfigUtils;
import io.unitycatalog.spark.auth.CredPropsUtil;
import io.unitycatalog.spark.utils.OptionsUtil;
import java.net.URI;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.hadoop.fs.Path;
import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
import org.apache.spark.sql.catalyst.catalog.CatalogUtils;
import org.apache.spark.sql.connector.catalog.Column;
import org.apache.spark.sql.connector.catalog.DelegatingCatalogExtension;
import org.apache.spark.sql.connector.catalog.Identifier;
import org.apache.spark.sql.connector.catalog.NamespaceChange;
import org.apache.spark.sql.connector.catalog.SupportsNamespaces;
import org.apache.spark.sql.connector.catalog.Table;
import org.apache.spark.sql.connector.catalog.TableCatalog;
import org.apache.spark.sql.connector.catalog.TableChange;
import org.apache.spark.sql.connector.expressions.Transform;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.util.CaseInsensitiveStringMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.sparkproject.guava.base.Preconditions;

/** A Spark catalog plugin to get/manage tables in Unity Catalog. */
public class UCSingleCatalog implements TableCatalog, SupportsNamespaces {
  private static final Logger LOG = LoggerFactory.getLogger(UCSingleCatalog.class);

  static final ThreadLocal<Boolean> LOAD_DELTA_CATALOG = ThreadLocal.withInitial(() -> true);
  static final ThreadLocal<Boolean> DELTA_CATALOG_LOADED = ThreadLocal.withInitial(() -> false);

  private URI uri = null;
  private TokenProvider tokenProvider = null;
  private boolean renewCredEnabled = false;
  private SchemasApi schemasApi = null;
  private TablesApi tablesApi = null;
  private TemporaryCredentialsApi tempCredApi = null;

  private volatile TableCatalog delegate = null;

  @Override
  public void initialize(String name, CaseInsensitiveStringMap options) {
    String urlStr = options.get(OptionsUtil.URI);
    Preconditions.checkArgument(
        urlStr != null, "uri must be specified for Unity Catalog '%s'", name);
    try {
      uri = new URI(urlStr);
    } catch (Exception e) {
      throw new RuntimeException("Invalid URI: " + urlStr, e);
    }
    tokenProvider = TokenProvider.create(AuthConfigUtils.buildAuthConfigs(options));
    renewCredEnabled =
        OptionsUtil.getBoolean(
            options,
            OptionsUtil.RENEW_CREDENTIAL_ENABLED,
            OptionsUtil.DEFAULT_RENEW_CREDENTIAL_ENABLED);

    // Initialize the ApiClient.
    ApiClient apiClient =
        ApiClientFactory.createApiClient(
            JitterDelayRetryPolicy.builder().build(), uri, tokenProvider);

    // Initialize the UCProxy.
    tempCredApi = new TemporaryCredentialsApi(apiClient);
    schemasApi = new SchemasApi(apiClient);
    tablesApi = new TablesApi(apiClient);
    UCProxy proxy =
        new UCProxy(uri, tokenProvider, renewCredEnabled, schemasApi, tablesApi, tempCredApi);
    proxy.initialize(name, options);

    // Initialize the delegate catalog.
    if (LOAD_DELTA_CATALOG.get()) {
      try {
        delegate =
            (TableCatalog)
                Class.forName("org.apache.spark.sql.delta.catalog.DeltaCatalog")
                    .getDeclaredConstructor()
                    .newInstance();
        ((DelegatingCatalogExtension) delegate).setDelegateCatalog(proxy);
        delegate.initialize(name, options);
        DELTA_CATALOG_LOADED.set(true);
      } catch (ClassNotFoundException e) {
        LOG.warn("DeltaCatalog is not available in the classpath", e);
        delegate = proxy;
      } catch (Exception e) {
        throw new RuntimeException("Failed to initialize DeltaCatalog", e);
      }
    } else {
      delegate = proxy;
    }
  }

  @Override
  public String name() {
    return delegate.name();
  }

  @Override
  public Identifier[] listTables(String[] namespace) throws NoSuchNamespaceException {
    return delegate.listTables(namespace);
  }

  @Override
  public Table loadTable(Identifier ident) throws NoSuchTableException {
    return delegate.loadTable(ident);
  }

  public Table loadTable(Identifier ident, String version) throws NoSuchTableException {
    return delegate.loadTable(ident, version);
  }

  public Table loadTable(Identifier ident, long timestamp) throws NoSuchTableException {
    return delegate.loadTable(ident, timestamp);
  }

  @Override
  public boolean tableExists(Identifier ident) {
    return delegate.tableExists(ident);
  }

  @Override
  public Table createTable(
      Identifier ident, Column[] columns, Transform[] partitions, Map<String, String> properties) {
    checkUnsupportedNestedNamespace(ident.namespace());
    boolean hasExternalClause = properties.containsKey(TableCatalog.PROP_EXTERNAL);
    boolean hasLocationClause = properties.containsKey(TableCatalog.PROP_LOCATION);
    if (hasExternalClause && !hasLocationClause) {
      throw new RuntimeException("Cannot create EXTERNAL TABLE without location.");
    }
    boolean isPathTable = ident.namespace().length == 1 && new Path(ident.name()).isAbsolute();

    // If both EXTERNAL and LOCATION are not specified in the CREATE TABLE command, and the table is
    // not a path table like parquet.`/file/path`, we generate the UC-managed table location here.
    if (!hasExternalClause && !hasLocationClause && !isPathTable) {
      // Check that caller shouldn't set some properties
      List<String> restrictedProps =
          Arrays.asList(
              UCTableProperties.UC_TABLE_ID_KEY,
              UCTableProperties.UC_TABLE_ID_KEY_OLD,
              TableCatalog.PROP_IS_MANAGED_LOCATION);
      for (String prop : restrictedProps) {
        if (properties.containsKey(prop)) {
          throw new RuntimeException("Cannot specify property '" + prop + "'.");
        }
      }
      // Setting the catalogManaged table feature is required for creating a managed table.
      if (!properties.containsKey(UCTableProperties.DELTA_CATALOG_MANAGED_KEY)
          && !properties.containsKey(UCTableProperties.DELTA_CATALOG_MANAGED_KEY_NEW)) {
        throw new RuntimeException(
            "Managed table creation requires table property '"
                + UCTableProperties.DELTA_CATALOG_MANAGED_KEY_NEW
                + "'='"
                + UCTableProperties.DELTA_CATALOG_MANAGED_VALUE
                + "' to be set.");
      }
      // Caller should not set these two table properties to values other than "supported". This is
      // the only documented value.
      List<String> catalogManagedKeys =
          Arrays.asList(
              UCTableProperties.DELTA_CATALOG_MANAGED_KEY,
              UCTableProperties.DELTA_CATALOG_MANAGED_KEY_NEW);
      for (String k : catalogManagedKeys) {
        String v = properties.get(k);
        if (v != null && !v.equals(UCTableProperties.DELTA_CATALOG_MANAGED_VALUE)) {
          throw new RuntimeException("Invalid property value '" + v + "' for '" + k + "'.");
        }
      }

      // Get staging table location and table id from UC
      CreateStagingTable createStagingTable =
          new CreateStagingTable()
              .catalogName(name())
              .schemaName(ident.namespace()[0])
              .name(ident.name());
      StagingTableInfo stagingTableInfo;
      try {
        stagingTableInfo = tablesApi.createStagingTable(createStagingTable);
      } catch (ApiException e) {
        throw new RuntimeException("Failed to create staging table", e);
      }
      String stagingLocation = stagingTableInfo.getStagingLocation();
      String stagingTableId = stagingTableInfo.getId();

      Map<String, String> newProps = new HashMap<>(properties);
      newProps.put(TableCatalog.PROP_LOCATION, stagingTableInfo.getStagingLocation());
      // Sets both the new and old table ID property while it's being renamed.
      newProps.put(UCTableProperties.UC_TABLE_ID_KEY, stagingTableInfo.getId());
      newProps.put(UCTableProperties.UC_TABLE_ID_KEY_OLD, stagingTableInfo.getId());
      // `PROP_IS_MANAGED_LOCATION` is used to indicate that the table location is not
      // user-specified but system-generated, which is exactly the case here.
      newProps.put(TableCatalog.PROP_IS_MANAGED_LOCATION, "true");

      TemporaryCredentials temporaryCredentials;
      try {
        temporaryCredentials =
            tempCredApi.generateTemporaryTableCredentials(
                new GenerateTemporaryTableCredential()
                    .tableId(stagingTableId)
                    .operation(TableOperation.READ_WRITE));
      } catch (ApiException e) {
        throw new RuntimeException("Failed to generate temporary credentials", e);
      }
      Map<String, String> credentialProps =
          CredPropsUtil.createTableCredProps(
              renewCredEnabled,
              CatalogUtils.stringToURI(stagingLocation).getScheme(),
              uri.toString(),
              tokenProvider,
              stagingTableId,
              TableOperation.READ_WRITE,
              temporaryCredentials);
      setCredentialProps(newProps, credentialProps);

      try {
        return delegate.createTable(ident, columns, partitions, newProps);
      } catch (org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
          | NoSuchNamespaceException e) {
        throw new RuntimeException("Failed to create table", e);
      }
    } else if (hasLocationClause) {
      String location = properties.get(TableCatalog.PROP_LOCATION);
      assert location != null;
      TemporaryCredentials cred;
      try {
        cred =
            tempCredApi.generateTemporaryPathCredentials(
                new GenerateTemporaryPathCredential()
                    .url(location)
                    .operation(PathOperation.PATH_CREATE_TABLE));
      } catch (ApiException e) {
        throw new RuntimeException("Failed to generate temporary path credentials", e);
      }
      Map<String, String> newProps = new HashMap<>(properties);

      Map<String, String> credentialProps =
          CredPropsUtil.createPathCredProps(
              renewCredEnabled,
              CatalogUtils.stringToURI(location).getScheme(),
              uri.toString(),
              tokenProvider,
              location,
              PathOperation.PATH_CREATE_TABLE,
              cred);

      setCredentialProps(newProps, credentialProps);
      try {
        return delegate.createTable(ident, columns, partitions, newProps);
      } catch (org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
          | NoSuchNamespaceException e) {
        throw new RuntimeException("Failed to create table", e);
      }
    } else {
      // TODO: for path-based tables, Spark should generate a location property using the qualified
      //       path string.
      try {
        return delegate.createTable(ident, columns, partitions, properties);
      } catch (org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException
          | NoSuchNamespaceException e) {
        throw new RuntimeException("Failed to create table", e);
      }
    }
  }

  @Override
  public Table createTable(
      Identifier ident, StructType schema, Transform[] partitions, Map<String, String> properties) {
    throw new UnsupportedOperationException("deprecated `createTable` should not be called");
  }

  @Override
  public Table alterTable(Identifier ident, TableChange... changes) {
    throw new UnsupportedOperationException("Altering a table is not supported yet");
  }

  @Override
  public boolean dropTable(Identifier ident) {
    return delegate.dropTable(ident);
  }

  @Override
  public void renameTable(Identifier oldIdent, Identifier newIdent) {
    throw new UnsupportedOperationException("Renaming a table is not supported yet");
  }

  @Override
  public String[][] listNamespaces() throws NoSuchNamespaceException {
    return ((DelegatingCatalogExtension) delegate).listNamespaces();
  }

  @Override
  public String[][] listNamespaces(String[] namespace) throws NoSuchNamespaceException {
    return ((DelegatingCatalogExtension) delegate).listNamespaces(namespace);
  }

  @Override
  public Map<String, String> loadNamespaceMetadata(String[] namespace)
      throws NoSuchNamespaceException {
    return ((DelegatingCatalogExtension) delegate).loadNamespaceMetadata(namespace);
  }

  @Override
  public void createNamespace(String[] namespace, Map<String, String> metadata) {
    try {
      ((DelegatingCatalogExtension) delegate).createNamespace(namespace, metadata);
    } catch (org.apache.spark.sql.catalyst.analysis.NamespaceAlreadyExistsException e) {
      throw new RuntimeException("Namespace already exists", e);
    }
  }

  @Override
  public void alterNamespace(String[] namespace, NamespaceChange... changes)
      throws NoSuchNamespaceException {
    ((DelegatingCatalogExtension) delegate).alterNamespace(namespace, changes);
  }

  @Override
  public boolean dropNamespace(String[] namespace, boolean cascade)
      throws NoSuchNamespaceException {
    try {
      return ((DelegatingCatalogExtension) delegate).dropNamespace(namespace, cascade);
    } catch (org.apache.spark.sql.catalyst.analysis.NonEmptyNamespaceException e) {
      throw new RuntimeException("Namespace is not empty", e);
    }
  }

  // Static methods from companion object

  static void setCredentialProps(Map<String, String> props, Map<String, String> credentialProps) {
    props.putAll(credentialProps);
    // TODO: Delta requires the options to be set twice in the properties, with and without the
    //       `option.` prefix. We should revisit this in Delta.
    credentialProps.forEach((k, v) -> props.put(TableCatalog.OPTION_PREFIX + k, v));
  }

  static void checkUnsupportedNestedNamespace(String[] namespace) {
    if (namespace.length > 1) {
      throw new RuntimeException(
          "Nested namespaces are not supported: " + String.join(".", namespace));
    }
  }

  /**
   * Constructs a fully qualified table name for Unity Catalog API calls.
   *
   * <p>This method creates a three-part name in the format `catalog.schema.table` by combining the
   * catalog name with the schema name (from the identifier's namespace) and table name. It is NOT
   * backtick quoted like what is usually used in SQL statements even if the names have special
   * characters like hyphens.
   *
   * <p>Example: catalogName=catalog, ident=(schema, table): it returns "catalog.schema.table"
   * catalogName=cata-log, ident=(sche-ma, ta-ble): returns "cata-log.sche-ma.ta-ble"
   * catalogName=catalog, ident=((schema1, schema2), table): throws ApiException(Nested namespace
   * not supported)
   *
   * @param catalogName the name of the catalog
   * @param ident the table identifier containing the namespace (schema) and table name
   * @return a fully qualified table name in the format "catalog.schema.table"
   * @throws ApiException if the identifier contains nested namespaces
   */
  static String fullTableNameForApi(String catalogName, Identifier ident) {
    checkUnsupportedNestedNamespace(ident.namespace());
    return catalogName + "." + ident.namespace()[0] + "." + ident.name();
  }
}
